Move list item checkbox to component

paulfalgout opened the pull request:
Shortcut Story ID: [sc-29967]

The checkbox re-rendering the whole item row caused all of the components to re-render.  Not terrible per-row, but in a multiselect situation, it really stacked up.  I _think_ we might still have a memory leak somewhere, but I'm still looking into that, and can measure better once this makes it to QA.

paulfalgout committed:
Move list item checkbox to component

Previously the checkbox rendered as part of the list row causing the entire row to re-render.  This was not preformant on long lists with large owner select lists.

paulfalgout committed:
Mult-select golfing

nmajor25 commented on src/js/views/patients/worklist/worklist_views.js:
```diff
@@ -206,15 +206,13 @@ const ListView = CollectionView.extend({
     const lastSelectedIndex = this.state.get('lastSelectedIndex');
 
     if (isShiftKeyPressed && lastSelectedIndex !== null && !isSelected) {
-      this.handleClickShiftMultiSelect(selectedIndex);
```
Nice catch. Better to only get `lastSelectedIndex` the one time.

nmajor25 commented on src/js/views/patients/shared/components/check_component.js:
```diff
@@ -0,0 +1,33 @@
+  viewOptions() {
+    const isSelected = this.getState('isSelected');
+    const template = isSelected ? hbs`{{fas "check-square"}}` : hbs`{{fal "square"}}`;
+    return {
```
_Nitpick alert_.

Add space between  the `const template` and the `return {` lines?

nmajor25 commented on src/js/views/patients/worklist/action_views.js:
```diff
@@ -48,8 +48,8 @@ const ActionItemView = View.extend({
-      'select:multiple': this.render,
-      'select:none': this.render,
+      'select:multiple': this.showCheck,
+      'select:none': this.showCheck,
```
After the bulk edit save is finished, the row items in the UI aren't updated.

To reproduce: Go to the worklist => select one or more list items => make a change (e.g. switch state from `To do` to `Doing` => hit `Save` button. When the bulk edit is finished, those list items aren't updated in the UI. And the list needs to be reloaded to see those changes (via a hard refresh of the app or leaving the worklist page and coming back).

Think it's related to this: https://github.com/RoundingWell/care-ops-frontend/blob/develop/src/js/apps/patients/worklist/worklist_app.js#L166.

Before this PR, `state.clearSelected()` was called when the bulk edit is finished, which triggers `select:none` and a `this.render()` on the list item.

nmajor25 replied:
Looks like the date/time are updated in the UI correctly. But the state and owner are not...? Now I'm confused...

paulfalgout replied:
Good catch on this one.  I'm a little surprised that no tests failed with this.  I'm assuming what I just pushed will need line coverage, but either way I plan on adding tests to cover what should have failed here.

nmajor25 replied:
From your last two commits, everything looks to be working well! Also was surprised the tests didn't catch it.

paulfalgout replied:
There's a bit of event loop because some of the custom saves don't check for change..  so it works kind of.. that's why there tests are currently failing. :-/

paulfalgout committed:
Rerender action/flow components on bulk model changes

paulfalgout commented on src/js/apps/patients/sidebar/bulk-edit-actions_app.js:
```diff
@@ -147,6 +147,7 @@ export default App.extend({
     });
   },
   onSubmit() {
+    this.modal.disableSubmit();
```
with this, the green save button was staying active while awaiting submission.. particularly noticeable on a patient flow page..and if someone hammered those, because it's parallel calls they could really stack up.  There might be more to do here, but this seems better than it was

nmajor25 replied:
Nice, that works well 👍 

One thought I had while testing this in my local sandbox. Should the `Cancel` option be changed to something else like `Clear`, `Clear Changes`, etc. (this sits next to the `Save` button in the sidebar)? Or just hide it once the `Save` button is clicked by the user?

Because once the `Save` button is clicked and the request is sent to the backend (where db updates are made), that request can't be cancelled. At that point, clicking `Cancel` just closes the sidebar, but doesn't really cancel anything.

paulfalgout replied:
Yeah maybe on the cancel thing.  Worth bringing up, but it'll be true everywhere.  The naming of the button is a UX decision.

paulfalgout committed:
Disable sidebar modal button on submit

Currently these buttons are reclickable.  We might need to look into a better way to do this, particularly for handling error states, but a overly disabled button is better than an overly enabled button.

paulfalgout committed:
Add simple test to ensure row is updating from bulk editing

paulfalgout commented:
@nmajor25 alright, so this was pulling the thread of a sweater..  This PR will fix the the multi-select issues, but it keeps/moves the issue over to bulk edit submitting.  This can be done, but not only does the data need to update, but potentially other state like whether or not the component should be disabled.  But now I have a dev sandbox so I can probably work on tracking down any leak better in the near future.
